### DISCUSSION TITLE
Runtime Class Creation & Manipulation ("Coremodding")

### DISCUSSION THREAD

**Discussion by Rongmario** [CleanroomMC/Cleanroom#87] [Draft]
### Summary

Improve "coremodding", which in this instance is primarily describing the context of transforming and possibly redefinition of classes at runtime.

### Goals

- Make the "coremodding" process less daunting & error-prone.
- Decide upon low level APIs to be included for aid.

### Non Goals

- Not to create new frameworks for bytecode manipulation.
- This CFP should not step into Bouncepad and transformer implementation territories.

### Motivation

Discussions happening on Discord every so often on how coremodding can be improved, and how Mixin and raw ASM may have a large learning curve for people that want to do basic class generation or transformations.

### Description

Forge's class transformation API is extremely primitive. In fact, it doesn't even exist. It all sits in the `LaunchWrapper` which is extremely legacy ~~garbage~~ baggage. And [Bouncepad](https://github.com/CleanroomMC/Bouncepad) is taking care of most of it. However, the development of it also rests on this CFP, as we need to come to an agreement of what we should do for coremodding.

First of all, no matter what libraries we include, we must make sure either of the points is achieved:

- The libraries should have sufficient and documented API for modders to use.
Or:
- We have to provide public-facing API that is sufficient and documented.

### Libraries:

Some of the public known libraries people have mentioned over on Discord can improve coremodding:

1. Burningwave Core
2. Javassist
3. Bytebuddy

(List is welcomed to be populated before this proposal is moved onto the next stage).

If possible, we should not go down the route of "js coremods" that modern Forge adopted.

### Dependencies

The above listed libraries (possibly).

### Guidelines

- [X] I have and will continue to adhere to the guidelines

**Comment by kappa-maintainer** [comment]
After two weeks using Javassist I can confirm this library is a must-have. Some tedious ASM work are just a few lines in Javassist. It's instrument method also provide some minimal @ Redirect and remap function.
Besides, the other two libraries are both designed for class generating, not transforming. BECL is a toy for educating (last time I check). We don't have many options here.
